rmf_traffic/src/rmf_traffic/schedule/RectifierInternal.hpp

```diff
@@ -29,9 +29,10 @@ class Rectifier::Implementation
 {
 public:
 
-  Participant::Implementation& participant;
+  std::weak_ptr<Participant::Implementation::Shared> participant;
```

Comment:
Am I right to say the reason for using a `weak_ptr` here is to avoid circular dependency with the `RectificationRequester` stored in `Participant::Implementation::Shared`?

Reply:
The relationships between these different classes is admittedly pretty convoluted. That convolution comes from the fact that we need to have a closed feedback loop to be robust to schedule update messages getting dropped on a network, but we also need to allow an arbitrary middleware to get plugged into two opposite points in that loop.

Whether or not there's a circular dependency between `Rectifier` and `Participant::Implementation::Shared` will depend on exactly how the arbitrary middleware plugin is implemented. In the case of [`DatabaseRectificationRequester`](https://github.com/open-rmf/rmf_traffic/blob/2af737c98f78a4029e4811dad00baa098f30116f/rmf_traffic/src/rmf_traffic/schedule/Rectifier.cpp#L57) and [`rmf_traffic_ros2::schedule::RectifierFactory`](https://github.com/open-rmf/rmf_ros2/blob/2cb313991db0be11147b250b8e4c85b3553c8e2e/rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Writer.cpp#L47) I think you're correct.

But even if there wasn't a risk of a circular dependency, the more critical motivation to use a `weak_ptr` here is this:
* A participant should not **have** to stay alive just because the rectifier for it is still alive.

If the rectifier is the last living thing that depends on the existence of the participant, then that implies that the user can no longer modify or update the participant in any way. It also implies that the user no longer cares about the existence of the participant. In that case, we should assume that the participant is not meant to exist anymore, and the rectifier should behave as though the participant has been unregistered.

Reply:
Thanks for the explanation 👍🏼